### PR TITLE
IGNITE-22403 CDC: Improve Kafka partition ranges assignment for appliers

### PR DESCRIPTION
**Pull request by maksaska**
Distribution for partitions from kafka topic over requested threads in `AbstractKafkaToIgniteCdcStreamer` (CDC Kafka to Ignite) is not uniform.

There might be cases with unequal load on threads, that reads data from kafka, which might result in overhead/bottleneck and slow workflow in general for the whole CDC with Kafka process. The improved version takes that into account. 

Initially, for uniform distribution to form, the streamer calculated the ratio between the number of partitions and the number of threads.It didn't take into account the modulus of the result, thus leading to underestimated number of partiotions one thread can take. The fixed version of the algorithm uses modulus to round up the underestimated number to pump up the load on the first threads. To make things work, the number of partitions taken by each thread is reevaulated for each thread, which allows to reduce the difference in parity and integral division in general. In other words, threads balance themself out while partitions are being distributed from the first thread to the last one.

### CODE REVIEW

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 1]
```diff
@@ -197,6 +188,34 @@ protected void runAppliers() {
...
     }
 
+    /**
+     * Creates Kafka partitions distribution for provided streamer configuration. Each element of the returning
+     * {@code List} is an {@link IgniteBiTuple}, which represents a kafka topic partition interval to be
+     * scanned by single thread.
+     * @param streamerCfg {@link KafkaToIgniteCdcStreamerConfiguration}.
+     * @return {@code List} of {@code IgniteBiTuple<Integer, Integer>}.
+     */
+    public static List<IgniteBiTuple<Integer, Integer>> getKafkaPartitionsDistribution(KafkaToIgniteCdcStreamerConfiguration streamerCfg) {
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:198]:
```suggestion
    public static List<IgniteBiTuple<Integer, Integer>> kafkaPartitions(KafkaToIgniteCdcStreamerConfiguration streamerCfg) {
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 2]
```diff
@@ -197,6 +188,34 @@ protected void runAppliers() {
...
+        List<IgniteBiTuple<Integer, Integer>> distr = new ArrayList<>();
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:199]:
```suggestion
        List<IgniteBiTuple<Integer, Integer>> parts = new ArrayList<>();
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 3]
```diff
@@ -197,6 +188,34 @@ protected void runAppliers() {
...
+
+            kafkaParts -= partPerApplier;
+            --threadCnt;
+
+            distr.add(new IgniteBiTuple<>(kafkaPartsFrom, kafkaPartsFrom + partPerApplier));
+
+            kafkaPartsFrom += partPerApplier;
+        }
+
+        return distr;
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:216]:
```suggestion
        return parts;
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 4]
```diff
@@ -153,26 +154,16 @@ protected void runAppliers() {
...
-
-        for (int i = 0; i < threadCnt; i++) {
-            int from = i * partPerApplier;
-            int to = (i + 1) * partPerApplier;
-
-            if (i == threadCnt - 1)
-                to = kafkaParts;
+        int counter = 0;
 
+        for (IgniteBiTuple<Integer, Integer> part : getKafkaPartitionsDistribution(streamerCfg)) {
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:159]:
```suggestion
        for (IgniteBiTuple<Integer, Integer> parts : kafkaPartitions(streamerCfg)) {
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 5]
```diff
@@ -153,26 +154,16 @@ protected void runAppliers() {
...
             KafkaToIgniteCdcStreamerApplier applier = new KafkaToIgniteCdcStreamerApplier(
                 () -> eventsApplier(),
                 log,
                 kafkaProps,
                 streamerCfg.getTopic(),
-                kafkaPartsFrom + from,
-                kafkaPartsFrom + to,
+                part.get1(),
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:165]:
```suggestion
                parts.get1(), // kafkaPartFrom
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 6]
```diff
@@ -153,26 +154,16 @@ protected void runAppliers() {
...
+                part.get2(),
```
Comment (shishkovilja) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java:166]:
```suggestion
                parts.get2(), // kafkaPartTo
```

**Review by shishkovilja — CHANGES_REQUESTED** [review]

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java (hunk 1 above)
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Done!

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java (hunk 3 above)
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Renamed it

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java (hunk 4 above)
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Done

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java (hunk 2 above)
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Done

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java (hunk 6 above)
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Done

**Review by shishkovilja — APPROVED** [review]

**Review by shishkovilja — CHANGES_REQUESTED** [review]
Please, fix my comments

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 7]
```diff
@@ -153,26 +154,16 @@ protected void runAppliers() {
...
-
-        int partPerApplier = kafkaParts / threadCnt;
-
-        for (int i = 0; i < threadCnt; i++) {
-            int from = i * partPerApplier;
-            int to = (i + 1) * partPerApplier;
-
-            if (i == threadCnt - 1)
-                to = kafkaParts;
+        int counter = 0;
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Renamed the variable accroding to the rules

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java [hunk 8]
```diff
@@ -181,7 +172,7 @@ protected void runAppliers() {
                 stopped
             );
 
-            addAndStart("applier-thread-" + i, applier);
+            addAndStart("applier-thread-" + counter++, applier);
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/AbstractKafkaToIgniteCdcStreamer.java]: Renamed the variable accroding to the rules

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+        testKafkaPartitionsCase(3, 1, 4, Arrays.asList(F.t(1, 2), F.t(2, 3), F.t(3, 4)));
+    }
+
+    /**
+     * Checks the topic partitions distribution over specified number of threads for specified parameters.
+     * @param threadCnt Thread count.
+     * @param result List of pairs defining partition ranges for each applier thread.
+     */
+    private void testKafkaPartitionsCase(int threadCnt, int partFrom, int partTo, List<IgniteBiTuple<Integer, Integer>> result) {
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Renamed the variable accroding to the rules

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        testKafkaPartitionsCase(3, 3, 11, Arrays.asList(F.t(3, 6), F.t(6, 9), F.t(9, 11)));
+        testKafkaPartitionsCase(4, 3, 11, Arrays.asList(F.t(3, 5), F.t(5, 7), F.t(7, 9), F.t(9, 11)));
+        testKafkaPartitionsCase(5, 3, 11, Arrays.asList(F.t(3, 5), F.t(5, 7), F.t(7, 9), F.t(9, 10), F.t(10, 11)));
+
+        testKafkaPartitionsCase(3, 1, 4, Arrays.asList(F.t(1, 2), F.t(2, 3), F.t(3, 4)));
+    }
+
+    /**
+     * Checks the topic partitions distribution over specified number of threads for specified parameters.
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Done!

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import org.junit.Test;
+
+import static org.apache.ignite.cdc.kafka.AbstractKafkaToIgniteCdcStreamer.kafkaPartitions;
+
+/** Tests the kafka topic partitions distribution over specified number of threads. */
+public class KafkaToIgnitePartitionDistributionTest extends GridCommonAbstractTest {
+    /** Streamer config. */
+    private final KafkaToIgniteCdcStreamerConfiguration streamerCfg = new KafkaToIgniteCdcStreamerConfiguration();
+
+    /** Checks the topic partitions distribution over specified number of threads. */
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Done!

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Arrays;
+import java.util.List;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.cdc.kafka.AbstractKafkaToIgniteCdcStreamer.kafkaPartitions;
+
+/** Tests the kafka topic partitions distribution over specified number of threads. */
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Changed the javadoc message

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        testKafkaPartitionsCase(3, 3, 11, Arrays.asList(F.t(3, 6), F.t(6, 9), F.t(9, 11)));
+        testKafkaPartitionsCase(4, 3, 11, Arrays.asList(F.t(3, 5), F.t(5, 7), F.t(7, 9), F.t(9, 11)));
+        testKafkaPartitionsCase(5, 3, 11, Arrays.asList(F.t(3, 5), F.t(5, 7), F.t(7, 9), F.t(9, 10), F.t(10, 11)));
+
+        testKafkaPartitionsCase(3, 1, 4, Arrays.asList(F.t(1, 2), F.t(2, 3), F.t(3, 4)));
+    }
+
+    /**
+     * Checks the topic partitions distribution over specified number of threads for specified parameters.
+     * @param threadCnt Thread count.
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Done!

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+     * @param result List of pairs defining partition ranges for each applier thread.
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Done!

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    private void testKafkaPartitionsCase(int threadCnt, int partFrom, int partTo, List<IgniteBiTuple<Integer, Integer>> result) {
+        streamerCfg.setThreadCount(threadCnt);
+        streamerCfg.setKafkaPartsFrom(partFrom);
+        streamerCfg.setKafkaPartsTo(partTo);
+
+        assertEquals(result, kafkaPartitions(streamerCfg));
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/kafka/KafkaToIgnitePartitionDistributionTest.java]: Done!

**Review by shishkovilja — APPROVED** [review]

**Review by NSAmelchev — APPROVED** [review]

**Pull request merged**